server/api/src/rsDataObjRead.cpp

```diff
@@ -55,40 +61,64 @@ applyRuleForPostProcForRead( rsComm_t *rsComm, bytesBuf_t *dataObjReadOutBBuf, c
 }
 
 int
-rsDataObjRead( rsComm_t *rsComm, openedDataObjInp_t *dataObjReadInp,
-               bytesBuf_t *dataObjReadOutBBuf ) {
-    int bytesRead;
-    int l1descInx = dataObjReadInp->l1descInx;
-
-    if ( l1descInx < 2 || l1descInx >= NUM_L1_DESC ) {
-        rodsLog( LOG_NOTICE,
-                 "rsDataObjRead: l1descInx %d out of range",
-                 l1descInx );
+rsDataObjRead(rsComm_t* rsComm,
+              openedDataObjInp_t* dataObjReadInp,
+              bytesBuf_t* dataObjReadOutBBuf)
+{
+    const int l1descInx = dataObjReadInp->l1descInx;
+
+    if (l1descInx < 2 || l1descInx >= NUM_L1_DESC) {
+        rodsLog(LOG_ERROR, "%s: l1descInx %d out of range", __func__, l1descInx);
         return SYS_FILE_DESC_OUT_OF_RANGE;
     }
-    if ( L1desc[l1descInx].inuseFlag != FD_INUSE ) {
+
+    auto& l1desc = L1desc[l1descInx];
+
+    if (l1desc.inuseFlag != FD_INUSE) {
         return BAD_INPUT_DESC_INDEX;
     }
 
-    if ( L1desc[l1descInx].remoteZoneHost != NULL ) {
-        /* cross zone operation */
-        dataObjReadInp->l1descInx = L1desc[l1descInx].remoteL1descInx;
-        bytesRead = rcDataObjRead( L1desc[l1descInx].remoteZoneHost->conn,
-                                   dataObjReadInp, dataObjReadOutBBuf );
+    if (l1desc.remoteZoneHost) {
+        // Cross zone operation.
+        dataObjReadInp->l1descInx = l1desc.remoteL1descInx;
+        const auto bytes_read = rcDataObjRead(l1desc.remoteZoneHost->conn, dataObjReadInp, dataObjReadOutBBuf);
         dataObjReadInp->l1descInx = l1descInx;
+        return bytes_read;
     }
-    else {
-        int i;
-        bytesRead = l3Read( rsComm, l1descInx, dataObjReadInp->len,
-                            dataObjReadOutBBuf );
-        i = applyRuleForPostProcForRead( rsComm, dataObjReadOutBBuf,
-                                         L1desc[l1descInx].dataObjInfo->objPath );
-        if ( i < 0 ) {
-            return i;
+
+    // Do not let the client read past the data size in the catalog if the replica
+    // was opened in read-only mode.
+    if (O_RDONLY == (l1desc.dataObjInp->openFlags & O_ACCMODE)) {
+        OpenedDataObjInp input{};
+        input.l1descInx = l1descInx;
+        input.whence = SEEK_CUR;
+
+        FileLseekOut* output{};
+
+        if (const auto ec = rsDataObjLseek(rsComm, &input, &output); ec < 0) {
+            rodsLog(LOG_ERROR, "%s: Could not retrieve the current file read position.", __func__, ec);
+            return ec;
         }
+
+        // If the file read position is greater than or equal to the data size,
+        // then return immediately.
+        if (output->offset >= l1desc.dataObjInfo->dataSize) {
+            return 0;
```

Comment:
is this correct?    not an error?    it's someone else's job to interpret this?

Reply:
This is the system telling the client that no bytes were read because they have reached the end of the file.

Reply:
or it's just saying it literally read zero bytes.